src/Storages/HDFS/HDFSCommon.cpp

```diff
@@ -0,0 +1,185 @@
+                    ErrorCodes::EXCESSIVE_ELEMENT_IN_CONFIG);
+            }
+
+            hadoop_security_kerberos_ticket_cache_path = config.getString(key_path);
+            // standard param - pass further
+        }
+
+        key_name = boost::replace_all_copy(key, "_", ".");
+
+        const auto & [k,v] = keep(key_name, config.getString(key_path));
+        hdfsBuilderConfSetStr(hdfs_builder, k.c_str(), v.c_str());
+    }
+}
+
+String HDFSBuilderWrapper::getKinitCmd()
+{
+    WriteBufferFromOwnString ss;
+
+    String cache_name =  hadoop_security_kerberos_ticket_cache_path.empty() ?
+        String() :
+        (String(" -c \"") + hadoop_security_kerberos_ticket_cache_path + "\"");
+
+    ss << hadoop_kerberos_kinit_command << cache_name <<
+        " -R -t \"" << hadoop_kerberos_keytab << "\" -k " << hadoop_kerberos_principal <<
+        "|| " << hadoop_kerberos_kinit_command << cache_name << " -t \"" <<
+        hadoop_kerberos_keytab << "\" -k " << hadoop_kerberos_principal;
+    return ss.str();
+}
+
+void HDFSBuilderWrapper::runKinit()
+{
+    String cmd = getKinitCmd();
+    LOG_DEBUG(&Poco::Logger::get("HDFSClient"), "running kinit: {}", cmd);
+
+    std::unique_lock<std::mutex> lck(kinit_mtx);
+
+    auto command = ShellCommand::execute(cmd);
+    auto status = command->tryWait();
+    if (status)
+    {
+        throw Exception("kinit failure: " + cmd, ErrorCodes::BAD_ARGUMENTS);
+    }
+}
+
+HDFSBuilderWrapper createHDFSBuilder(const String & uri_str, const Context & context)
+{
+    const Poco::URI uri(uri_str);
+    const auto & host = uri.getHost();
+    auto port = uri.getPort();
+    const String path = "//";
+    if (host.empty())
+        throw Exception("Illegal HDFS URI: " + uri.toString(), ErrorCodes::BAD_ARGUMENTS);
+
+    HDFSBuilderWrapper builder;
+    if (builder.get() == nullptr)
+        throw Exception("Unable to create builder to connect to HDFS: " +
+            uri.toString() + " " + String(hdfsGetLastError()),
+            ErrorCodes::NETWORK_ERROR);
+
+    hdfsBuilderConfSetStr(builder.get(), "input.read.timeout", "60000"); // 1 min
+    hdfsBuilderConfSetStr(builder.get(), "input.write.timeout", "60000"); // 1 min
+    hdfsBuilderConfSetStr(builder.get(), "input.connect.timeout", "60000"); // 1 min
+
+    String user_info = uri.getUserInfo();
+    String user;
+    if (!user_info.empty() && user_info.front() != ':')
+    {
+        size_t delim_pos = user_info.find(':');
+        if (delim_pos != String::npos)
+            user = user_info.substr(0, delim_pos);
+        else
+            user = user_info;
+
+        hdfsBuilderSetUserName(builder.get(), user.c_str());
+    }
+    hdfsBuilderSetNameNode(builder.get(), host.c_str());
+    if (port != 0)
+    {
+        hdfsBuilderSetNameNodePort(builder.get(), port);
+    }
+
+    // const auto & config = context.getGlobalContext().getConfigRef();
+    const auto & config = context.getConfigRef();
```

Comment:
We just need a config from context. Why are we passing context everywhere?

Reply:
Actual reasons are (1) because context is passed everywhere in kafka and (2) because I considered a dedicated thread for kinit, but decided to keep it simple.
No one is a good excuse and I'll switch to config.
Thanks.

Reply:
fixed